packages/kit/src/runtime/client/router.js

```diff
@@ -109,12 +109,16 @@ export class Router {
 		const handle_mousemove = (event) => {
 			clearTimeout(mousemove_timeout);
 			mousemove_timeout = setTimeout(() => {
-				trigger_prefetch(event);
+				// ensure event.composedPath() in find_anchor(event) returns an array of EventTarget objects
+				event.target?.dispatchEvent(
+					new CustomEvent('sveltekit:trigger_prefetch', { bubbles: true })
```

Comment:
I wonder if something like `sveltekit:mousestop` might be a better name for this? What do you guys think?

Reply:
Either works well for me, but it does looks nicer when used in the `addEventListener` call below.

Reply:
I lean toward `sveltekit:trigger_prefetch` as well.

Reply:
My instinct is to look for an alternative solution altogether — it seems weird to dispatch a custom event just to indirectly call our own function, no?

Reply:
Actually, I realised there's a deeper problem here. The `composedPath` stuff was [introduced](https://github.com/sveltejs/kit/pull/2769) because recursively checking `parentNode` to find an `<a>` element doesn't work if the `<a>` in question is in shadow DOM.

The solution here dispatches a custom event on `e.target`, but the `composedPath` of the custom event _won't_ include the shadow `<a>` in certain cases. [Demo here](https://stackblitz.com/edit/vite-gdntva) — basically, if you have a custom element with an `<a>` in the shadow DOM like this...

```js
customElements.define(
  'fancy-a',
  class extends HTMLElement {
    constructor() {
      super();

      let a = document.createElement('a');
      a.href = this.getAttribute('href');
      a.innerHTML = '<slot></slot>';

      let shadowRoot = this.attachShadow({ mode: 'open' });
      shadowRoot.append(a);
    }
  }
);
```

...then whether or not the `<a>` will be included in the `composedPath` depends on whether the original event happened to happen within an element in light DOM inside the custom element.

I'm not exactly sure how to solve this, or if it's even possible. We might have to just say that prefetching won't work with shadow DOM.

Reply:
Ok, it looks like shadow roots have an `elementsFromPoint` method, like `document`, which means we could solve it [thusly](https://stackblitz.com/edit/vite-5cmvnt):

```js
function find_anchor(x, y, root = document) {
  for (const element of root.elementsFromPoint(x, y)) {
    if (element.tagName === 'A') return element;
    if (element.shadowRoot && element.shadowRoot !== root) {
      const found = find_anchor(x, y, element.shadowRoot);
      if (found) return found;
    }
  }
}
```

This does run the risk of false positives (e.g. an element is on top of the `<a>`, blocking clicks) but I guess nonsense like this is just the price we pay for supporting web components. Perhaps there's a neat way to use `composedPath` for `touchstart` and only use `elementsFromPoint` for the mouse events.

Reply:
That does look a bit risky and potentially more resource intensive. I wonder if it would be fine to not support `sveltekit:prefetch` in web components at all, the usecase would be rare. And also that normal link clicks should already work in web components.

Reply:
I would be totally fine dropping support for prefetch with web components

Reply:
I'm on the pragmatic side here. I don't think that dispatching a custom element with an explanation why this is needed is bad nor does it clutter the code.

Reply:
The biggest problem with the custom event is that Rich showed it doesn't work all the times. I share the sentiment from @bluwy that the issue is becoming rather cumbersome.

Reply:
Very well, if we're okay only having partial support for shadow DOM then we can merge this — thanks. Very annoying that the existence of shadow DOM forces us to jump through these ridiculous hoops in the first place!